File: buildFlamechart.m
function handles = buildFlamechart(ax, callSequence, options)
% BUILDFLAMECHART Render a flamechart visualization of the call sequence.
%
%   handles = buildFlamechart(ax, callSequence)
%   handles = buildFlamechart(ax, callSequence, options)
%
%   Inputs:
%     ax           - Target axes handle
%     callSequence - struct array from parseProfilerPhases (enter/exit events with depth)
%     options      - (optional) struct with fields:
%       .highlightName - function name to highlight (default: '')
%       .colormap      - Nx3 colormap matrix (default: lines(256))
%
%   Output:
%     handles - struct array with fields: rect, text, funcName, seqIndex
%               One entry per "enter" event that has a matching "exit".

    if nargin < 3
        options = struct();
    end

    highlightName = getOpt(options, 'highlightName', '');
    cmap          = getOpt(options, 'colormap', lines(256));

    cla(ax);
    hold(ax, 'on');

    nEntries = numel(callSequence);
    if nEntries == 0
        title(ax, 'No data to display');
        return;
    end

    % We need to pair enter/exit events to draw rectangles.
    % Build rectangles for each "enter" by finding its matching "exit".
    % Use a stack-based approach.
    maxDepth = 0;
    rects = struct('x', {}, 'width', {}, 'depth', {}, 'funcName', {}, 'fileName', {}, 'seqIndex', {});
    rectIdx = 0;

    % Map sequence position to an x-coordinate.
    % We track the x position as a running counter of enter events.
    enterCount = 0;
    % Stack: each entry is [enterXPos, depth, seqIndex, funcName, fileName]
    stack = {};

    for k = 1:nEntries
        entry = callSequence(k);
        if strcmp(entry.event, 'enter')
            enterCount = enterCount + 1;
            stack{end+1} = struct('xStart', enterCount, 'depth', entry.depth, ...
                                  'seqIndex', entry.index, 'funcName', entry.funcName, ...
                                  'fileName', entry.fileName); %#ok<AGROW>
            if entry.depth > maxDepth
                maxDepth = entry.depth;
            end
        elseif strcmp(entry.event, 'exit') && ~isempty(stack)
            % Pop the most recent matching entry from the stack
            % Find the topmost stack entry with the same funcName
            popIdx = numel(stack);
            for s = numel(stack):-1:1
                if strcmp(stack{s}.funcName, entry.funcName)
                    popIdx = s;
                    break;
                end
            end
            info = stack{popIdx};
            stack(popIdx) = [];

            rectIdx = rectIdx + 1;
            rects(rectIdx).x        = info.xStart;
            rects(rectIdx).width    = enterCount - info.xStart + 1;
            rects(rectIdx).depth    = info.depth;
            rects(rectIdx).funcName = info.funcName;
            rects(rectIdx).fileName = info.fileName;
            rects(rectIdx).seqIndex = info.seqIndex;
        end
    end

    % Also flush any remaining stack entries (unmatched enters)
    for s = 1:numel(stack)
        info = stack{s};
        rectIdx = rectIdx + 1;
        rects(rectIdx).x        = info.xStart;
        rects(rectIdx).width    = enterCount - info.xStart + 1;
        rects(rectIdx).depth    = info.depth;
        rects(rectIdx).funcName = info.funcName;
        rects(rectIdx).fileName = info.fileName;
        rects(rectIdx).seqIndex = info.seqIndex;
    end

    if rectIdx == 0
        title(ax, 'No rectangles to draw');
        return;
    end

    % Assign colors by function name
    uniqueNames = unique({rects.funcName});
    nColors = size(cmap, 1);
    colorMap = containers.Map();
    for k = 1:numel(uniqueNames)
        cidx = mod(k - 1, nColors) + 1;
        colorMap(uniqueNames{k}) = cmap(cidx, :);
    end

    % Draw rectangles
    handles = struct('rect', {}, 'text', {}, 'funcName', {}, 'seqIndex', {}, 'fileName', {});
    barHeight = 0.85;

    for k = 1:rectIdx
        r = rects(k);
        xPos = r.x - 0.5;
        yPos = r.depth - 1;
        w = r.width;
        h = barHeight;

        faceColor = colorMap(r.funcName);

        % Highlight if requested
        if ~isempty(highlightName) && contains(r.funcName, highlightName, 'IgnoreCase', true)
            edgeColor = [1 0 0];
            lineWidth = 2.5;
        else
            edgeColor = [0.3 0.3 0.3];
            lineWidth = 0.5;
        end

        rectH = rectangle(ax, 'Position', [xPos, yPos, w, h], ...
            'FaceColor', faceColor, 'EdgeColor', edgeColor, ...
            'LineWidth', lineWidth, 'Curvature', [0.05, 0.05]);

        % Add text label if rectangle is wide enough
        textH = [];
        if w > 2
            % Truncate label if needed
            label = r.funcName;
            textH = text(ax, xPos + w/2, yPos + h/2, label, ...
                'HorizontalAlignment', 'center', 'VerticalAlignment', 'middle', ...
                'FontSize', 7, 'Interpreter', 'none', 'Clipping', 'on');
        end

        idx = numel(handles) + 1;
        handles(idx).rect     = rectH;
        handles(idx).text     = textH;
        handles(idx).funcName = r.funcName;
        handles(idx).seqIndex = r.seqIndex;
        handles(idx).fileName = r.fileName;

        % Store metadata for click callback
        rectH.UserData = struct('funcName', r.funcName, 'fileName', r.fileName, ...
                                'seqIndex', r.seqIndex, 'depth', r.depth);
    end

    hold(ax, 'off');

    % Configure axes
    ax.YDir = 'reverse';
    ax.XLim = [0, enterCount + 1];
    ax.YLim = [0, maxDepth + 1];
    xlabel(ax, 'Call Sequence Position');
    ylabel(ax, 'Call Depth');
    title(ax, 'Flamechart');
    ax.TickLabelInterpreter = 'none';

    % Enable interactive navigation
    zoom(ancestor(ax, 'figure'), 'on');
end

function val = getOpt(opts, fieldName, default)
    if isfield(opts, fieldName)
        val = opts.(fieldName);
    else
        val = default;
    end
end

File: Python_Version/build_flamechart.py
"""Build a flamechart visualization from a call sequence.

Draws stacked rectangles on a matplotlib Axes where x = sequence position
and y = call depth. Supports highlighting by function name and click callbacks.
"""

from dataclasses import dataclass, field
from typing import Optional, Callable

import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.colors as mcolors
import numpy as np

from .parse_profiler_phases import CallEvent


@dataclass
class FlamechartRect:
    """Metadata for one rectangle in the flamechart."""
    x: float
    width: float
    depth: int
    func_name: str
    file_name: str
    seq_index: int
    patch: Optional[mpatches.FancyBboxPatch] = None
    text: Optional[plt.Text] = None


def build_flamechart(
    ax: plt.Axes,
    call_sequence: list[CallEvent],
    *,
    highlight_name: str = "",
    colormap: Optional[np.ndarray] = None,
    on_click: Optional[Callable[[FlamechartRect], None]] = None,
) -> list[FlamechartRect]:
    """Render a flamechart on the given axes.

    Args:
        ax: Matplotlib axes to draw on.
        call_sequence: List of CallEvent from parse_profiler_phases.
        highlight_name: Function name substring to highlight in red.
        colormap: Nx3 array of RGB colors (0-1). Defaults to tab20.
        on_click: Callback invoked when a rectangle is clicked.

    Returns:
        List of FlamechartRect with references to the drawn patches.
    """
    ax.clear()

    if not call_sequence:
        ax.set_title("No data to display")
        return []

    if colormap is None:
        cmap = plt.cm.tab20
        colormap = np.array([cmap(i / 20) [:3] for i in range(20)])

    # Pair enter/exit events using a stack to build rectangles
    @dataclass
    class _StackEntry:
        x_start: int
        depth: int
        seq_index: int
        func_name: str
        file_name: str

    stack: list[_StackEntry] = []
    rects: list[FlamechartRect] = []
    enter_count = 0
    max_depth = 0

    for entry in call_sequence:
        if entry.event == "enter":
            enter_count += 1
            stack.append(_StackEntry(
                x_start=enter_count,
                depth=entry.depth,
                seq_index=entry.index,
                func_name=entry.func_name,
                file_name=entry.file_name,
            ))
            max_depth = max(max_depth, entry.depth)
        elif entry.event == "exit" and stack:
            # Pop the topmost matching entry
            pop_idx = len(stack) - 1
            for s in range(len(stack) - 1, -1, -1):
                if stack[s].func_name == entry.func_name:
                    pop_idx = s
                    break
            info = stack.pop(pop_idx)
            rects.append(FlamechartRect(
                x=info.x_start,
                width=enter_count - info.x_start + 1,
                depth=info.depth,
                func_name=info.func_name,
                file_name=info.file_name,
                seq_index=info.seq_index,
            ))

    # Flush remaining stack entries (unmatched enters)
    for info in stack:
        rects.append(FlamechartRect(
            x=info.x_start,
            width=enter_count - info.x_start + 1,
            depth=info.depth,
            func_name=info.func_name,
            file_name=info.file_name,
            seq_index=info.seq_index,
        ))

    if not rects:
        ax.set_title("No rectangles to draw")
        return []

    # Assign colors by function name
    unique_names = sorted(set(r.func_name for r in rects))
    n_colors = len(colormap)
    color_map = {
        name: colormap[i % n_colors] for i, name in enumerate(unique_names)
    }

    bar_height = 0.85
    highlight_lower = highlight_name.lower() if highlight_name else ""

    for r in rects:
        x_pos = r.x - 0.5
        y_pos = r.depth - 1
        w = r.width
        h = bar_height

        face_color = color_map[r.func_name]

        if highlight_lower and highlight_lower in r.func_name.lower():
            edge_color = "red"
            line_width = 2.5
        else:
            edge_color = (0.3, 0.3, 0.3)
            line_width = 0.5

        patch = mpatches.FancyBboxPatch(
            (x_pos, y_pos), w, h,
            boxstyle="round,pad=0.02",
            facecolor=face_color,
            edgecolor=edge_color,
            linewidth=line_width,
        )
        ax.add_patch(patch)
        r.patch = patch

        # Add text label if rectangle is wide enough
        if w > 2:
            txt = ax.text(
                x_pos + w / 2, y_pos + h / 2, r.func_name,
                ha="center", va="center", fontsize=6,
                clip_on=True,
            )
            r.text = txt

    # Set up click handler
    if on_click:
        def _on_pick(event):
            for r in rects:
                if r.patch is not None and r.patch.contains(event)[0]:
                    on_click(r)
                    break

        ax.figure.canvas.mpl_connect("button_press_event", _on_pick)

    ax.set_xlim(0, enter_count + 1)
    ax.set_ylim(-0.5, max_depth + 1)
    ax.invert_yaxis()
    ax.set_xlabel("Call Sequence Position")
    ax.set_ylabel("Call Depth")
    ax.set_title("Flamechart")

    return rects


def highlight_rects_by_name(
    rects: list[FlamechartRect], func_name: str
) -> None:
    """Highlight all rectangles matching func_name, dim the rest."""
    for r in rects:
        if r.patch is None:
            continue
        if r.func_name.lower() == func_name.lower():
            r.patch.set_edgecolor("red")
            r.patch.set_linewidth(2.5)
        else:
            r.patch.set_edgecolor((0.3, 0.3, 0.3))
            r.patch.set_linewidth(0.5)
